src/main/java/com/microsoft/azure/vmagent/builders/AzureVMCloudBuilder.java

```diff
@@ -0,0 +1,121 @@
+package com.microsoft.azure.vmagent.builders;
+
+import com.microsoft.azure.vmagent.AzureVMAgentTemplate;
+import com.microsoft.azure.vmagent.AzureVMCloud;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class AzureVMCloudBuilder {
+
+    private String cloudName;
+
+    private String azureCredentialsId;
+
+    private String maxVirtualMachinesLimit;
+
+    private String deploymentTimeout;
+
+    private String resourceGroupReferenceType;
+
+    private String newResourceGroupName;
+
+    private String existingResourceGroupName;
+
+    private List<AzureVMAgentTemplate> vmTemplates;
+
+    public AzureVMCloudBuilder() {
+        maxVirtualMachinesLimit = "10";
+        deploymentTimeout = "1200";
+        resourceGroupReferenceType = "new";
+        vmTemplates = new ArrayList<>();
+    }
+
+    //CHECKSTYLE:OFF
+    public AzureVMCloudBuilder withCloudName(String cloudName) {
+        this.cloudName = cloudName;
+        return this;
+    }
+
+    public AzureVMCloudBuilder withAzureCredentialsId(String azureCredentialsId) {
+        this.azureCredentialsId = azureCredentialsId;
+        return this;
+    }
+
+    public AzureVMCloudBuilder withMaxVirtualMachinesLimit(String maxVirtualMachinesLimit) {
+        this.maxVirtualMachinesLimit = maxVirtualMachinesLimit;
+        return this;
+    }
+
+    public AzureVMCloudBuilder withDeploymentTimeout(String deploymentTimeout) {
+        this.deploymentTimeout = deploymentTimeout;
+        return this;
+    }
+
+    public AzureVMCloudBuilder withNewResourceGroupName(String resourceGroupName) {
+        this.resourceGroupReferenceType = "new";
+        this.newResourceGroupName = resourceGroupName;
+        return this;
+    }
+
+    public AzureVMCloudBuilder withExistingResourceGroupName(String resourceGroupName) {
+        this.resourceGroupReferenceType = "existing";
+        this.existingResourceGroupName = resourceGroupName;
+        return this;
+    }
+
+    public AzureVMCloudBuilder withTemplates(List<AzureVMAgentTemplate> templates) {
+        this.vmTemplates.clear();
+        this.vmTemplates.addAll(templates);
+        return this;
+    }
+
+    public AzureVMCloudBuilder addToTemplates(List<AzureVMAgentTemplate> templates) {
+        this.vmTemplates.addAll(templates);
+        return this;
+    }
+
+    public AzureVMCloudBuilder addToTemplates(AzureVMAgentTemplate... templates) {
+        for (AzureVMAgentTemplate template : templates) {
+            this.vmTemplates.add(template);
+        }
+        return this;
+    }
+
+    public AzureVMTemplateNested addNewTemplate() {
```

Comment:
Will it be useful it have a method that copies from an existing template?

Reply:
In the current model, it will cause some confusing. `AdvancedImage`/`BasicImage` is a complete instance, so user can't copy a template from existing one but only change a part of arguments in `AdvancedImage`/`BasicImage`. Or we need to change the model to make `AdvancedImage`/`BasicImage` as the same level as `Template`, but it can't bring the same experience as the current model.